### PR TITLE
Rework testing conventions gradle plugin

### PR DESCRIPTION
**Pull request by breskeby**
This PR reworks the testing conventions precommit plugin. This plugin now:
- is compatible with yaml, java rest tests and internalClusterTest (aka different sourceSets per test type)
- enforces test base class and simple naming conventions (as it did before)
- adds one check task per test sourceSet
- uses the worker api to improve task execution parallelism and encapsulation
- is gradle configuration cache compatible  

This also ports the TestingConventions integration testing to Spock and removes the build-tools-internal/test kit folder that is not required anymore. We also add some common logic for testing java related gradle plugins. 
We will apply further cleanup on other tests within our test suite in a dedicated follow up cleanup

### CODE REVIEW

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/TestingConventionsCheckTask.java [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+    @Inject
+    public abstract WorkerExecutor getWorkerExecutor();
+
+    public void baseClass(String qualifiedClassname) {
+        getBaseClasses().add(qualifiedClassname);
+    }
+
+    @TaskAction
+    void validate() {
+        WorkQueue workQueue = getWorkerExecutor().classLoaderIsolation(spec -> spec.getClasspath().from(getClasspath()));
```
Comment (breskeby) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/TestingConventionsCheckTask.java:82]: For now we keep it in the process but at one point we should revisit the general memory settings for our build as more stuff moved out to dedicated workers

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/TestingConventionsPrecommitPlugin.java [hunk 2]
```diff
@@ -8,41 +8,79 @@
...
+            setupTaskPerSourceSet(project, sourceSet, t -> {
+                t.getSuffix().convention("Tests");
+                t.getBaseClasses().convention(List.of("org.apache.lucene.tests.util.LuceneTestCase"));
             });
-        return testingConventions;
+        });
+
+        project.getPlugins().withType(InternalYamlRestTestPlugin.class, yamlRestTestPlugin -> {
+            NamedDomainObjectProvider<SourceSet> sourceSet = sourceSets.named(InternalYamlRestTestPlugin.SOURCE_SET_NAME);
+            setupTaskPerSourceSet(project, sourceSet, t -> {
```
Comment (breskeby) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/TestingConventionsPrecommitPlugin.java:47]: At the moment we apply the same base class defaults for yaml and java rest tests. We might introduce a dedicated base class for each of them at one point.

**Comment by breskeby** [comment]
Screenshot 2022-06-13 at 11 59 34

**Comment by breskeby** [comment]
@mark-vieira @pugnascotia would you mind having a look / review at this?

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/TestingConventionsCheckTask.java [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+                && method.getReturnType().equals(Void.TYPE);
+        }
+
+        private static boolean isAnnotated(Method method, Class<?> annotation) {
+            return List.of(method.getAnnotations())
+                .stream()
+                .anyMatch(presentAnnotation -> annotation.isAssignableFrom(presentAnnotation.getClass()));
+        }
+
+        private List<String> walkPathAndLoadClasses(File testRoot) {
```
Comment (mark-vieira) [inline build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/TestingConventionsCheckTask.java:215]: Couldn't we simplify this by using `FileTree.visit()` and leveraging `RelativePath`?

**Review by mark-vieira — APPROVED** [review]
Couple minor comments but otherwise this looks great.

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/TestingConventionsCheckTask.java [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+                Class<?> junitTest = loadClassWithoutInitializing("org.junit.Assert", clazz.getClassLoader());
+                if (junitTest.isAssignableFrom(clazz)) {
+                    Logging.getLogger(TestingConventionsCheckWorkAction.class)
+                        .debug("{} is a test because it extends {}", clazz.getName(), junitTest.getName());
+                    return true;
+                }
+
+                Class<?> junitAnnotation = loadClassWithoutInitializing("org.junit.Test", clazz.getClassLoader());
+                for (Method method : clazz.getMethods()) {
+                    if (matchesTestMethodNamingConvention(method)) {
```
[the remark this answers is not in the document]
Comment (breskeby) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/TestingConventionsCheckTask.java:179]: it's more about detecting those as wrongly declared tests. We should not have those.

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/TestingConventionsPrecommitPlugin.java [hunk 5]
```diff
@@ -8,41 +8,96 @@
...
+        });
+
+        // Create a convenience task for all checks (this does not conflict with extension, as it has higher priority in DSL):
+        return tasks.register(TESTING_CONVENTIONS_TASK_NAME, task -> {
+            task.setDescription("Runs all testing conventions checks.");
+            task.dependsOn(tasks.withType(TestingConventionsCheckTask.class));
+        });
     }
+
+    private void setupTaskPerSourceSet(
```
[the remark this answers is not in the document]
Comment (breskeby) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/TestingConventionsPrecommitPlugin.java:87]: fixed

File: build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/TestingConventionsCheckTask.java (hunk 3 above)
Comment (breskeby) [reply build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/precommit/TestingConventionsCheckTask.java]: just copied this over from the original class. yeah I simplified this now

**Comment by mark-vieira** [comment]
@breskeby were we intending to backport this? This is a pretty significant change and it's already making backporting stuff to older branches conflict.

**Pull request merged**